### PR TITLE
conversions

### PR DESCRIPTION
**Pull request by blacktwin**
Support Listing Conversions #131

### CODE REVIEW

**Comment by blacktwin** [comment]
Looks like playlists/1111 is not universal. I'll need to find a way to find this key used for background processing.

**Comment by andrey-yantsen** [comment]
Seems like you have to call `/playlists?type=42` to get a correct id. For me the response is following:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<MediaContainer size="1">
<Playlist ratingKey="1" key="/playlists/1/items" guid="com.plexapp.agents.none://<redacted>" type="playlist" title="Background Processing List" summary="" playlistType="backgroundProcessing" addedAt="1506580160">
</Playlist>
</MediaContainer>
```

Furthermore, to get an actual conversion items (because in the playlist — it's not the media itself, or, at least, it may be not only media) you have to query for `playQueues?type=42` and then get a playQueue with the received id.

**Comment by blacktwin** [comment]
Ha, and I thought type42 would have not been universal. I should have flipped it. Thanks.

**Comment by blacktwin** [comment]
As far as the request/issue goes, I believe this returns the list of conversions. Like in the WebUI there is no connection from the Conversion menu and items to the actual media item. Unless you were inferring that we can @andrey-yantsen ?

**Comment by andrey-yantsen** [comment]
As far as I understand the API:
* `/playlists?type=42` returns a playlist where you can get all the `Optimised Versions` you've created, it's not about Conversions list
* `/playQueues?type=42` returns a playqueue with `Conversions` list, for me the response is following:
    ```xml
    <?xml version="1.0" encoding="UTF-8"?>
    <MediaContainer size="1">
    <PlayQueue id="1" clientIdentifier="<redacted>" totalItemsCount="7" type="playlist" shuffled="0" />
    </MediaContainer>
    ```
* `/playQueues/1` (where `1` is the id from previous request) returns an actual media list for Conversions page
    ```xml
    <?xml version="1.0" encoding="UTF-8"?>
    <MediaContainer size="7" identifier="com.plexapp.plugins.library" mediaTagPrefix="/system/bundle/media/flags/" mediaTagVersion="1570658413" playQueueID="1" playQueueSelectedItemID="125893" playQueueSelectedItemOffset="0" playQueueSelectedMetadataItemID="10378" playQueueShuffled="0" playQueueTotalCount="7" playQueueVersion="14456">
    <Video generatorID="2120" generatorType="42" playQueueItemID="125893" playlistID="1" target="Pre-sync Lenovo 4mbps" userID="1" username="<redacted>" ratingKey="10378" key="/library/metadata/10378" skipParent="1" parentRatingKey="10375" grandparentRatingKey="10374" guid="com.plexapp.agents.thetvdb://361857/1/3?lang=en" parentGuid="com.plexapp.agents.thetvdb://361857/1?lang=en" grandparentGuid="com.plexapp.agents.thetvdb://361857?lang=en" librarySectionTitle="TV Shows" librarySectionID="2" librarySectionKey="/library/sections/2" type="episode" title="It&#39;s All My Fault" grandparentKey="/library/metadata/10374" parentKey="/library/metadata/10375" grandparentTitle="Dead to Me" parentTitle="Season 1" contentRating="TV-PG" summary="" index="3" parentIndex="1" year="2019" thumb="/library/metadata/10378/thumb/1570312055" art="/library/metadata/10374/art/1570312058" parentThumb="/library/metadata/10375/thumb/1570312058" grandparentThumb="/library/metadata/10374/thumb/1570312058" grandparentArt="/library/metadata/10374/art/1570312058" grandparentTheme="/library/metadata/10374/theme/1570312058" originallyAvailableAt="2019-05-03" addedAt="1570309208" updatedAt="1570312055">
    </Video>
    <Video clientIdentifier="<redacted>" generatorID="2140" generatorType="1" playQueueItemID="125894" playlistID="1" syncItemId="39511457" syncListId="33899332" userID="1" username="<redacted>" ratingKey="9431" key="/library/metadata/9431" guid="com.plexapp.agents.imdb://tt9657544?lang=en" librarySectionTitle="Stand-ups" librarySectionID="14" librarySectionKey="/library/sections/14" studio="Comedy Dynamics" type="movie" title="Jim Gaffigan: Quality Time" summary="" rating="7.3" year="2019" thumb="/library/metadata/9431/thumb/1570577291" art="/library/metadata/9431/art/1570577291" duration="4500000" originallyAvailableAt="2019-08-16" addedAt="1567829396" updatedAt="1570577291">
    </Video>
    <Video clientIdentifier="<redacted>" generatorID="2140" generatorType="1" playQueueItemID="125896" playlistID="1" syncItemId="39511457" syncListId="33899332" userID="1" username="<redacted>" ratingKey="9511" key="/library/metadata/9511" guid="com.plexapp.agents.imdb://tt9169592?lang=en" librarySectionTitle="Stand-ups" librarySectionID="14" librarySectionKey="/library/sections/14" studio="Push It Productions" type="movie" title="Wanda Sykes: Not Normal" summary="" rating="6.9" year="2019" thumb="/library/metadata/9511/thumb/1570577296" art="/library/metadata/9511/art/1570577296" duration="3960000" originallyAvailableAt="2019-05-21" addedAt="1567920656" updatedAt="1570577296">
    </Video>
    <Video clientIdentifier="<redacted>" generatorID="2140" generatorType="1" playQueueItemID="125897" playlistID="1" syncItemId="39511457" syncListId="33899332" userID="1" username="<redacted>" ratingKey="9744" key="/library/metadata/9744" guid="com.plexapp.agents.imdb://tt9426212?lang=en" librarySectionTitle="Stand-ups" librarySectionID="14" librarySectionKey="/library/sections/14" type="movie" title="Gabriel Iglesias: One Show Fits All" summary="" rating="6.8" year="2019" thumb="/library/metadata/9744/thumb/1570577301" art="/library/metadata/9744/art/1570577301" duration="5400000" originallyAvailableAt="2019-01-03" addedAt="1568655377" updatedAt="1570577301">
    </Video>
    <Video clientIdentifier="<redacted>" generatorID="2140" generatorType="1" playQueueItemID="125898" playlistID="1" syncItemId="39511457" syncListId="33899332" userID="1" username="<redacted>" ratingKey="9711" key="/library/metadata/9711" guid="com.plexapp.agents.imdb://tt7157574?lang=en" librarySectionTitle="Stand-ups" librarySectionID="14" librarySectionKey="/library/sections/14" studio="Tenth Planet Productions" type="movie" title="Ellen DeGeneres: Relatable" summary="" rating="6.8" year="2018" tagline="She won&#8217;t be using her daytime voice." thumb="/library/metadata/9711/thumb/1570577299" art="/library/metadata/9711/art/1570577299" duration="4080000" originallyAvailableAt="2018-12-18" addedAt="1568411986" updatedAt="1570577299">
    </Video>
    <Video generatorID="2120" generatorType="42" playQueueItemID="125899" playlistID="1" target="Pre-sync Lenovo 4mbps" userID="1" username="<redacted>" ratingKey="10379" key="/library/metadata/10379" skipParent="1" parentRatingKey="10375" grandparentRatingKey="10374" guid="com.plexapp.agents.thetvdb://361857/1/4?lang=en" parentGuid="com.plexapp.agents.thetvdb://361857/1?lang=en" grandparentGuid="com.plexapp.agents.thetvdb://361857?lang=en" librarySectionTitle="TV Shows" librarySectionID="2" librarySectionKey="/library/sections/2" type="episode" title="I Can&#39;t Go Back" grandparentKey="/library/metadata/10374" parentKey="/library/metadata/10375" grandparentTitle="Dead to Me" parentTitle="Season 1" contentRating="TV-PG" summary="" index="4" parentIndex="1" year="2019" thumb="/library/metadata/10379/thumb/1570312055" art="/library/metadata/10374/art/1570312058" parentThumb="/library/metadata/10375/thumb/1570312058" grandparentThumb="/library/metadata/10374/thumb/1570312058" grandparentArt="/library/metadata/10374/art/1570312058" grandparentTheme="/library/metadata/10374/theme/1570312058" originallyAvailableAt="2019-05-03" addedAt="1570309208" updatedAt="1570312055">
    </Video>
    <Video generatorID="2120" generatorType="42" playQueueItemID="125900" playlistID="1" target="Pre-sync Lenovo 4mbps" userID="1" username="<redacted>" ratingKey="10380" key="/library/metadata/10380" skipParent="1" parentRatingKey="10375" grandparentRatingKey="10374" guid="com.plexapp.agents.thetvdb://361857/1/5?lang=en" parentGuid="com.plexapp.agents.thetvdb://361857/1?lang=en" grandparentGuid="com.plexapp.agents.thetvdb://361857?lang=en" librarySectionTitle="TV Shows" librarySectionID="2" librarySectionKey="/library/sections/2" type="episode" title="I&#39;ve Gotta Get Away" grandparentKey="/library/metadata/10374" parentKey="/library/metadata/10375" grandparentTitle="Dead to Me" parentTitle="Season 1" contentRating="TV-PG" summary="" index="5" parentIndex="1" year="2019" thumb="/library/metadata/10380/thumb/1570312055" art="/library/metadata/10374/art/1570312058" parentThumb="/library/metadata/10375/thumb/1570312058" grandparentThumb="/library/metadata/10374/thumb/1570312058" grandparentArt="/library/metadata/10374/art/1570312058" grandparentTheme="/library/metadata/10374/theme/1570312058" originallyAvailableAt="2019-05-03" addedAt="1570309208" updatedAt="1570312055">
    </Video>
    </MediaContainer>
    ```
    Items with `clientIdentifier`, `syncItemId` and `syncListId` are conversions for Mobile Sync, others — Optimised Versions. Probably there is some kind of pagination involved, or just a hardcoded limit, because not all items are being returned (you can see the total amount of items in `playQueueTotalCount` field of MediaContainer in this request, or in `totalItemsCount` in previous one).

**Comment by blacktwin** [comment]
Ahh. I wasn't understanding your first mention of `playQueues`. I'll look into it further. Thanks again.

**Comment by blacktwin** [comment]
So conversions are active jobs being processed. Here is my current output for `playQueues?type=42`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<MediaContainer size="1">
<PlayQueue id="1" clientIdentifier="<redacted>" totalItemsCount="-1" type="playlist" shuffled="0" />
</MediaContainer>
```

I guess this PR is incorrectly labeled as conversions when it's really pulling Optimized items. I should probably delete this PR and start over with the correct labeling and function naming. oof.

**Comment by andrey-yantsen** [comment]
Are you sure sure on the `Conversion` class? Shouldn't we be using regular Media subclasses for it?

**Comment by blacktwin** [comment]
I did look at that and there are a quite a few differences and new additional attributes. Unless you're suggesting I add the new additional attributes to the MediaObject instead? I'm open to suggestions and corrections.

**Comment by blacktwin** [comment]
Just wanting to mention that this PR is touching upon #130 now. So far, this PR is returning lists of Active Conversions (if in existence) and Optimized Items. As this PR has already grown past it's original intended issue, I think merging with just returning these lists might be a good idea. Merging once everyone has approved of the methods, obviously. Adding additional actions for start, pausing, resuming, and removing optimized items would be better in a new PR.

**Comment by blacktwin** [comment]
@andrey-yantsen were you referring to something like this?
```python
class Conversion(Media):
    """ Represents a Conversion item. """
    TAG = 'Media'
    TYPE = 'conversion'
    METADATA_TYPE = 'conversion'
```
and removing any duplicate attributes that exist in the higher level Media object?

**Comment by andrey-yantsen** [comment]
@blacktwin not exactly: I was thinking about using current `Video`/`Audio` objects instead of introducing a new `Conversion` object. But, honestly, I didn't thought it through properly, and have no idea whether it's really possible.

File: plexapi/server.py [hunk 1]
```diff
@@ -363,6 +364,30 @@ def playlist(self, title):
...
+        return items
+
+    def conversions(self):
+        """ Returns list of all :class:`~plexapi.media.Conversion` objects connected to server. """
+        items = []
+
+        backgroundProcessing = self.query('/playQueues?type=42')
+        for elem in backgroundProcessing:
+            key = elem.attrib.get('id')
+            for elem in self.query('/playQueues/%s' % key):
```
Comment (Hellowlol) [inline plexapi/server.py]: fetchItem

File: plexapi/server.py [hunk 2]
```diff
@@ -363,6 +364,30 @@ def playlist(self, title):
...
         return self.fetchItem('/playlists', title=title)
 
+    def optimizedItems(self):
+        """ Returns list of all :class:`~plexapi.media.Optimized` objects connected to server. """
+        items = []
+
+        backgroundProcessing = self.query('/playlists?type=42')
+        for elem in backgroundProcessing:
+            key = elem.attrib.get('key')
+            for elem in self.query(key):
```
[the remark this answers is not in the document]
Comment (blacktwin) [reply plexapi/server.py]: If we can assume that `/playQueues/1` is the default queue for conversions then
```python
return self.fetchItems('/playQueues/1')
```

or
```python
backgroundProcessing = self.query('/playQueues?type=42')
key = backgroundProcessing[0].attrib.get('id')
return self.fetchItems('/playQueues/%s' % key)
```
Either would return a partial `Video` object and not the `Conversion` object that I've created. So some information is gained and some is lost (video vs conversion object). The conversion object is true to the xml that is retrieved. Unless there is another way to force fetchItems to use the Conversion object, to which I'm struggling to find.

File: plexapi/server.py (hunk 1 above)
Comment (blacktwin) [reply plexapi/server.py]: If we can assume that `/playQueues/1` is the default queue for conversions then
```python
return self.fetchItems('/playQueues/1')
```

or
```python
backgroundProcessing = self.query('/playQueues?type=42')
key = backgroundProcessing[0].attrib.get('id')
return self.fetchItems('/playQueues/%s' % key)
```
Either would return a partial `Video` object and not the `Conversion` object that I've created. So some information is gained and some is lost (video vs conversion object). The conversion object is true to the xml that is retrieved. Unless there is another way to force fetchItems to use the Conversion object, to which I'm struggling to find.

Comment (Hellowlol) [reply plexapi/server.py]: The playqueue needs a rewamp, we are missing move, and add. I think i need  this for bw_plex so im gonna send a pr for that later. As for the fetchItems i think its a cls parameter that you can use.

**Comment by blacktwin** [comment]
@Hellowlol are we waiting for the update to the `playQueue`?

**Comment by Hellowlol** [comment]
No i can do that later.

**Comment by blacktwin** [comment]
@Hellowlol finally figured out my issue with using fetchItem(s). Should be good to go now.

**Comment by blacktwin** [comment]
As mentioned before I'd like to merge this PR with the scope of only listing Conversion and Optimized items. Once merged I'll start on adding methods for removing, pausing, creating, etc.

**Comment by Hellowlol** [comment]
I dont have time to review this. If you think its good and your tests pass that good enough for me.

**Comment by blacktwin** [comment]
No pytests yet as there are no methods for creating an optimized version yet. Once I add that functionality I'll be sure to add a pytest for it.

**Comment by blacktwin** [comment]
@pkkid Any input?

**Comment by Hellowlol** [comment]
Skimmed quick and seems fine.

**Comment by Hellowlol** [comment]
Add tests when you add the other methods.

**Pull request merged**